plugins/workspace-minimap/src/focus_mode.ts

```diff
@@ -0,0 +1,186 @@
+      if (parentSvg.firstChild) {
+        parentSvg.insertBefore(this.svgGroup, parentSvg.firstChild);
+      } else {
+        parentSvg.appendChild(this.svgGroup);
+      }
+
+      window.addEventListener('resize', () => void this.update());
+      this.onChangeWrapper = this.onChange.bind(this);
+      this.primaryWorkspace.addChangeListener(this.onChangeWrapper);
+
+      this.update();
+      this.initialized = true;
+    }
+
+
+    /**
+     * Disposes of the focus mode.
+     * Unlinks from all DOM elements and remove all event listeners
+     * to prevent memory leaks.
+     */
+    dispose() {
+      if (this.onChangeWrapper) {
+        this.primaryWorkspace.removeChangeListener(this.onChangeWrapper);
+      }
+      if (this.svgGroup) {
+        Blockly.utils.dom.removeNode(this.svgGroup);
+      }
+      this.svgGroup = null;
+      this.rect = null;
+      this.background = null;
+      this.initialized = false;
+    }
+
+
+    /**
+     * Handles events triggered on the primary workspace.
+     * @param event The event.
+     */
+    private onChange(event: Blockly.Events.Abstract): void {
+      if (blockEvents.has(event.type)) {
+        this.update();
+      }
+    }
+
+
+    /**
+     * Positions and sizes the highlight on the minimap
+     * based on the primary workspace.
+     */
+    private update(): void {
+      // Get the metrics.
+      const primaryMetrics = this.primaryWorkspace.getMetricsManager();
+      const minimapMetrics = this.minimapWorkspace.getMetricsManager();
+
+      const primaryView = primaryMetrics.getViewMetrics();
+      const primaryContent = primaryMetrics.getContentMetrics();
+      const minimapContent = minimapMetrics.getContentMetrics();
+      const minimapSvg = minimapMetrics.getSvgMetrics();
+
+      // Get the workscape to pixel scale on the minimap.
+      const scale = minimapContent.width /
+        minimapMetrics.getContentMetrics(true).width;
+
+      // Get the viewport size on a minimap scale.
+      const width = primaryView.width * scale;
+      const height = primaryView.height * scale;
+
+      // Get the viewport position in relation to the content.
+      let left = (primaryView.left - primaryContent.left) * scale;
+      let top = (primaryView.top - primaryContent.top) * scale;
+
+      // Account for the padding outside the content on the minimap.
+      left += (minimapSvg.width - minimapContent.width) / 2;
+      top += (minimapSvg.height - minimapContent.height) / 2;
+
+      // Set the svg attributes.
+      this.rect.setAttribute('transform',
+          'translate(' + left + ',' + top + ')');
+      this.rect.setAttribute('width', width.toString());
+      this.rect.setAttribute('height', height.toString());
+    }
+
+    /**
+     * Returns whether focus mode is initialized or not.
+     * @returns True if focus mode is initialized else false.
+     */
+    isEnabled(): boolean {
+      return this.initialized;
+    }
```

Comment:
You probably want to add something similar to the minimap as well (e.g. `isFocusEnabled`) because the focus mode is only accessible by subclasses. So if you just have a minimap, you can't tell if its focus mode is enabled or not atm.

Reply:
Yea I plan on doing that, however I was leaving it for a separate PR.

Reply:
Sweet! sounds good =)